contracts/Staking.sol

```diff
@@ -64,36 +70,62 @@ contract Staking {
         }
 
         // update the next epoch pool size
-        Pool storage pNextEpoch = poolSize[tokenAddress][nextEpoch];
+        Pool storage pNextEpoch = poolSize[tokenAddress][currentEpoch + 1];
         pNextEpoch.size = token.balanceOf(address(this));
         pNextEpoch.set = true;
 
         Checkpoint[] storage checkpoints = balanceCheckpoints[msg.sender][tokenAddress];
 
+        uint256 balanceBefore = getEpochUserBalance(msg.sender, tokenAddress, currentEpoch);
+
         // if there's no checkpoint yet, it means the user didn't have any activity
         // we want to store checkpoints both for the current epoch and next epoch because
         // if a user does a withdraw, the current epoch can also be modified and
         // we don't want to insert another checkpoint in the middle of the array as that could be expensive
         if (checkpoints.length == 0) {
-            checkpoints.push(Checkpoint(currentEpoch, 0));
-            checkpoints.push(Checkpoint(nextEpoch, balances[msg.sender][tokenAddress]));
+            checkpoints.push(Checkpoint(currentEpoch, currentMultiplier, 0, amount));
+
+            // next epoch => multiplier is 1, epoch deposits is 0
+            checkpoints.push(Checkpoint(currentEpoch + 1, BASE_MULTIPLIER, amount, 0));
```

Comment:
shouldn't amount be actually balances[msg.sender][tokenAddress] ?

Reply:
They're the same value in this case (first deposit ever).